template/base/validator.py

```diff
@@ -0,0 +1,177 @@
+from abc import ABC, abstractmethod
+from traceback import print_exception
+
+from template.utils.misc import ttl_get_block
+from template.utils.sync import sync, check_registered
+from template.utils.config import check_config, add_args, config
+
+
+class BaseValidatorNeuron(ABC):
+    """
+    Base class for Bittensor validators.
+    """
+
+    @classmethod
+    def check_config(cls, config: "bt.Config"):
+        check_config(cls, config)
+
+    @classmethod
+    def add_args(cls, parser):
+        add_args(cls, parser)
+
+    @classmethod
+    def config(cls):
+        return config(cls)
+
+    subtensor: "bt.subtensor"
+    wallet: "bt.wallet"
+    metagraph: "bt.metagraph"
+
+    def __init__(self, config=None):
+        base_config = copy.deepcopy(config or BaseValidatorNeuron.config())
+        self.config = self.config()
+        self.check_config(self.config)
+        self.config.merge(base_config)
+
+        # If a gpu is required, set the device to cuda:N (e.g. cuda:0)
+        self.device = self.config.neuron.device
+
+        # Log the configuration for reference.
+        bt.logging.info(self.config)
+
+        # Step 4: Build Bittensor validator objects
+        # These are core Bittensor classes to interact with the network.
+        bt.logging.info("Setting up bittensor objects.")
+
+        # The wallet holds the cryptographic key pairs for the validator.
+        self.wallet = bt.wallet(config=config)
+        bt.logging.info(f"Wallet: {self.wallet}")
+
+        # The subtensor is our connection to the Bittensor blockchain.
+        self.subtensor = bt.subtensor(config=self.config)
+        bt.logging.info(f"Subtensor: {self.subtensor}")
+
+        # Dendrite is the RPC client; it lets us send messages to other nodes (axons) in the network.
+        self.dendrite = bt.dendrite(wallet=self.wallet)
+        bt.logging.info(f"Dendrite: {self.dendrite}")
+
+        # The metagraph holds the state of the network, letting us know about other validators and miners.
+        self.metagraph = self.subtensor.metagraph(self.config.netuid)
+        bt.logging.info(f"Metagraph: {self.metagraph}")
+
+        # Save a copy of the hotkeys to local memory.
+        self.hotkeys = copy.deepcopy(self.metagraph.hotkeys)
+
+        # Set up logging with the provided configuration and directory.
+        bt.logging(config=self.config, logging_dir=self.config.full_path)
+        bt.logging.info(
+            f"Running validator for subnet: {self.config.netuid} on network: {self.subtensor.chain_endpoint}"
+        )
+
+        # Ensure we're registered on the subnet first.
+        check_registered(self)
+
+        self.uid = self.metagraph.hotkeys.index(self.wallet.hotkey.ss58_address)
+        bt.logging.info(f"Running validator on uid: {self.uid}")
+
+        # Set up initial scoring weights for validation
+        bt.logging.info("Building validation weights.")
+        self.scores = torch.ones_like(self.metagraph.S, dtype=torch.float32)
```

Comment:
Any reason why the scores start at one rather than zero?

Reply:
The primary reason is lack of coffee. Thank you.